File: api/waitlist/statistics.py
from functools import cached_property
from typing import Dict, TypeVar
import datetime

from flask import Blueprint, g
from sqlalchemy.sql import func

from . import auth
from .webutil import ViewReturn
from .data.database import FleetActivity, FitHistory, Fitting
from .data.dbutil import from_unixtime, yearmonth
from .data.evedb import name_of

bp = Blueprint("stats", __name__)

T = TypeVar("T")  # pylint: disable=invalid-name
N = TypeVar("N", int, float)  # pylint: disable=invalid-name


def _filter_into_other(
    source_data: Dict[str, Dict[str, N]],
    threshold: float = 0.01,
    other: str = "Other",
) -> Dict[str, Dict[str, N]]:
    sums: Dict[str, float] = {}
    for x_value in source_data.values():
        for key, value in x_value.items():
            if key not in sums:
                sums[key] = 0
            sums[key] += value

    total = sum(sums.values())
    top_n = {item[0] for item in sums.items() if item[1] > threshold * total}

    result: Dict[str, Dict[str, N]] = {}
    for x_key, x_value in source_data.items():
        result[x_key] = {}
        for key, value in x_value.items():
            if key in top_n:
                result[x_key][key] = value
            else:
                if not other in result[x_key]:
                    result[x_key][other] = 0
                result[x_key][other] += value

    return result


def _filter_into_other_1d(
    source_data: Dict[str, N], threshold: float = 0.01, other: str = "Other"
) -> Dict[str, N]:
    total = sum(source_data.values())
    result: Dict[str, N] = {}
    for key, value in source_data.items():
        if value >= threshold * total:
            result[key] = value
        else:
            if other not in result:
                result[other] = value
            else:
                result[other] += value
    return result


class StatisticsFetcher:
    @cached_property
    def beginning_of_time(  # Used for memoization. pylint: disable=no-self-use
        self,
    ) -> str:
        for (year_month,) in g.db.query(
            yearmonth(func.min(FitHistory.logged_at))
        ).all():
            return str(year_month)
        raise Exception("When did time begin?!")

    @cached_property
    def now_minus_28_days(  # Used for memoization. pylint: disable=no-self-use
        self,
    ) -> datetime.datetime:
        return datetime.datetime.utcnow() - datetime.timedelta(days=28)

    def _fill_missing_months(
        self, source_data: Dict[str, T], on_missing: T
    ) -> Dict[str, T]:
        begin_year, begin_month = map(int, self.beginning_of_time.split("-", 1))
        current_datetime = datetime.datetime.utcnow()

        iter_year = begin_year
        iter_month = begin_month

        result_data: Dict[str, T] = {}
        while iter_year < current_datetime.month or iter_month < current_datetime.month:
            iter_month += 1
            if iter_month > 12:
                iter_month = 1
                iter_year += 1
            year_month = "%04d-%02d" % (iter_year, iter_month)
            result_data[year_month] = source_data.get(year_month, on_missing)
        return result_data

    @cached_property
    def fleet_seconds_by_character_by_month(self) -> Dict[str, Dict[int, int]]:
        result: Dict[str, Dict[int, int]] = {}
        for year_month, character_id, fleet_seconds in (
            g.db.query(
                yearmonth(from_unixtime(FleetActivity.first_seen)),
                FleetActivity.character_id,
                func.sum(FleetActivity.last_seen - FleetActivity.first_seen),
            )
            .group_by(
                yearmonth(from_unixtime(FleetActivity.first_seen)),
                FleetActivity.character_id,
            )
            .all()
        ):
            result.setdefault(str(year_month), {})[character_id] = int(fleet_seconds)
        return self._fill_missing_months(result, {})

    @cached_property
    def fleet_seconds_by_hull_by_month(self) -> Dict[str, Dict[int, int]]:
        result: Dict[str, Dict[int, int]] = {}
        for year_month, hull, fleet_seconds in (
            g.db.query(
                yearmonth(from_unixtime(FleetActivity.first_seen)),
                FleetActivity.hull,
                func.sum(FleetActivity.last_seen - FleetActivity.first_seen),
            )
            .group_by(
                yearmonth(from_unixtime(FleetActivity.first_seen)),
                FleetActivity.hull,
            )
            .all()
        ):
            result.setdefault(str(year_month), {})[hull] = int(fleet_seconds)
        return self._fill_missing_months(result, {})

    @cached_property
    def xes_by_hull_by_month(self) -> Dict[str, Dict[int, int]]:
        result: Dict[str, Dict[int, int]] = {}
        for year_month, hull, count in (
            g.db.query(
                yearmonth(FitHistory.logged_at),
                Fitting.hull,
                func.count(func.distinct(FitHistory.character_id)),
            )
            .join(Fitting)
            .filter(Fitting.id == FitHistory.fit_id)
            .group_by(
                yearmonth(FitHistory.logged_at),
                Fitting.hull,
            )
            .all()
        ):
            result.setdefault(str(year_month), {})[hull] = int(count)
        return self._fill_missing_months(result, {})

    @cached_property
    def fleet_seconds_by_month(self) -> Dict[str, int]:
        result: Dict[str, int] = {}
        for year_month, fleet_seconds in (
            g.db.query(
                yearmonth(from_unixtime(FleetActivity.first_seen)),
                func.sum(FleetActivity.last_seen - FleetActivity.first_seen),
            )
            .group_by(yearmonth(from_unixtime(FleetActivity.first_seen)))
            .all()
        ):
            result[year_month] = int(fleet_seconds)
        return self._fill_missing_months(result, 0)

    @cached_property
    def pilots_by_month(self) -> Dict[str, int]:
        result: Dict[str, int] = {}
        for year_month, pilots in (
            g.db.query(
                yearmonth(from_unixtime(FleetActivity.first_seen)),
                func.count(func.distinct(FleetActivity.character_id)),
            )
            .group_by(yearmonth(from_unixtime(FleetActivity.first_seen)))
            .all()
        ):
            result[year_month] = int(pilots)
        return self._fill_missing_months(result, 0)

    @cached_property
    def xes_by_hull_28d(self) -> Dict[int, int]:
        result: Dict[int, int] = {}
        for hull, count in (
            g.db.query(
                Fitting.hull,
                func.count(func.distinct(FitHistory.character_id)),
            )
            .join(Fitting)
            .filter(
                Fitting.id == FitHistory.fit_id,
                FitHistory.logged_at > self.now_minus_28_days,
            )
            .group_by(
                Fitting.hull,
            )
            .all()
        ):
            result[hull] = int(count)
        return result

    @cached_property
    def fleet_seconds_by_hull_28d(self) -> Dict[int, int]:
        result: Dict[int, int] = {}
        for hull, fleet_seconds in (
            g.db.query(
                FleetActivity.hull,
                func.sum(FleetActivity.last_seen - FleetActivity.first_seen),
            )
            .group_by(FleetActivity.hull)
            .filter(FleetActivity.first_seen > self.now_minus_28_days.timestamp())
        ).all():
            result[hull] = int(fleet_seconds)
        return result


def _map_dict_keys_to_name(source_data: Dict[int, T]) -> Dict[str, T]:
    result: Dict[str, T] = {}
    for key, value in source_data.items():
        result[name_of(key)] = value
    return result


def build_fleet_seconds_by_hull_by_month(
    fetcher: StatisticsFetcher,
) -> Dict[str, Dict[str, int]]:
    return _filter_into_other(
        {
            year_month: _map_dict_keys_to_name(raw_data)
            for year_month, raw_data in fetcher.fleet_seconds_by_hull_by_month.items()
        }
    )


def build_xes_by_hull_by_month(fetcher: StatisticsFetcher) -> Dict[str, Dict[str, int]]:
    return _filter_into_other(
        {
            year_month: _map_dict_keys_to_name(raw_data)
            for year_month, raw_data in fetcher.xes_by_hull_by_month.items()
        }
    )


def build_time_spent_in_fleet_by_month(
    fetcher: StatisticsFetcher,
) -> Dict[str, Dict[str, int]]:
    result: Dict[str, Dict[str, int]] = {}
    buckets = ["a. <1h", "b. 1-5h", "c. 5-15h", "d. 15-40h", "e. 40h+"]
    for (
        year_month,
        character_data,
    ) in fetcher.fleet_seconds_by_character_by_month.items():
        result[year_month] = {bucket: 0 for bucket in buckets}
        for _character_id, fleet_seconds in character_data.items():
            if fleet_seconds < (1 * 3600):
                bucket = "a. <1h"
            elif fleet_seconds < (5 * 3600):
                bucket = "b. 1-5h"
            elif fleet_seconds < (15 * 3600):
                bucket = "c. 5-15h"
            elif fleet_seconds < (40 * 3600):
                bucket = "d. 15-40h"
            else:
                bucket = "e. 40h+"
            result[year_month][bucket] += 1
    return result


def build_xes_by_hull_28d(fetcher: StatisticsFetcher) -> Dict[str, int]:
    return _filter_into_other_1d(_map_dict_keys_to_name(fetcher.xes_by_hull_28d))


def build_fleet_seconds_by_hull_28d(fetcher: StatisticsFetcher) -> Dict[str, int]:
    return _filter_into_other_1d(
        _map_dict_keys_to_name(fetcher.fleet_seconds_by_hull_28d)
    )


def build_x_vs_time_by_hull_28d(
    fetcher: StatisticsFetcher,
) -> Dict[str, Dict[str, float]]:
    fleet_time_by_hull = fetcher.fleet_seconds_by_hull_28d
    xes_by_hull = fetcher.xes_by_hull_28d

    sum_fleet_time = sum(fleet_time_by_hull.values())
    sum_xes = sum(xes_by_hull.values())

    if sum_fleet_time == 0 or sum_xes == 0:
        # Avoid division by zero
        return {}

    listed_hulls = {
        item[0]
        for item in fleet_time_by_hull.items()
        if item[1] >= sum_fleet_time * 0.01  # 1% threshold
    }

    result: Dict[str, Dict[str, float]] = {}
    for hull, fleet_seconds in fleet_time_by_hull.items():
        bucket = name_of(hull) if hull in listed_hulls else "Other"
        result.setdefault(bucket, {"Time": 0, "X": 0})["Time"] += (
            fleet_seconds / sum_fleet_time
        )

    for hull, count in xes_by_hull.items():
        bucket = name_of(hull) if hull in listed_hulls else "Other"
        result.setdefault(bucket, {"Time": 0, "X": 0})["X"] += count / sum_xes

    return result


@bp.route("/api/stats")
@auth.login_required
@auth.require_permission("stats-view")
def stats() -> ViewReturn:
    fetcher = StatisticsFetcher()

    return {
        "beginning_of_time": str(fetcher.beginning_of_time),
        "fleet_seconds_by_hull_by_month": build_fleet_seconds_by_hull_by_month(fetcher),
        "xes_by_hull_by_month": build_xes_by_hull_by_month(fetcher),
        "fleet_seconds_by_month": dict(fetcher.fleet_seconds_by_month),
        "pilots_by_month": dict(fetcher.pilots_by_month),
        "time_spent_in_fleet_by_month": build_time_spent_in_fleet_by_month(fetcher),
        "xes_by_hull_28d": build_xes_by_hull_28d(fetcher),
        "fleet_seconds_by_hull_28d": build_fleet_seconds_by_hull_28d(fetcher),
        "x_vs_time_by_hull_28d": build_x_vs_time_by_hull_28d(fetcher),
    }

File: backend/src/routes/statistics.rs
use std::collections::BTreeMap;

use eve_data_core::{TypeDB, TypeError, TypeID};
use rocket::serde::json::Json;

use crate::{app::Application, core::auth::AuthenticatedAccount, util::madness::Madness};

use serde::Serialize;

macro_rules! from_unixtime {
    ( $a:expr ) => {
        concat!("from_unixtime(", $a, ")")
    };
}

macro_rules! year_month {
    ( $a:expr ) => {
        concat!("date_format(", $a, ", '%Y-%m')")
    };
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Clone, Copy)]
struct YearMonth(pub i32, pub i32);

impl std::fmt::Display for YearMonth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:04}-{:02}", self.0, self.1)
    }
}

impl YearMonth {
    fn parse(input: &str) -> YearMonth {
        let mut pieces = input.split('-');
        let first = pieces
            .next()
            .expect("First piece always there")
            .parse()
            .expect("Should parse");
        let second = pieces
            .next()
            .expect("Second piece better be there")
            .parse()
            .expect("Should parse");
        YearMonth(first, second)
    }
}

impl Serialize for YearMonth {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let formatted = format!("{}", self);
        serializer.serialize_str(&formatted)
    }
}

fn filter_into_other_1d(source: BTreeMap<String, f64>, threshold: f64) -> BTreeMap<String, f64> {
    let sum: f64 = source.values().sum();

    let mut other = 0.;
    let mut result = BTreeMap::new();
    for (name, value) in source {
        if value / sum > threshold {
            result.insert(name, value);
        } else {
            other += value;
        }
    }
    result.insert("Other".to_string(), other);

    result
}

fn filter_into_other_2d<T: Ord>(
    source: BTreeMap<T, BTreeMap<String, f64>>,
    threshold: f64,
) -> BTreeMap<T, BTreeMap<String, f64>> {
    let mut sums = BTreeMap::new();
    let mut total = 0.;

    for series in source.values() {
        for (name, &value) in series {
            *sums.entry(name.to_string()).or_insert(0.) += value;
            total += value;
        }
    }

    let mut result = BTreeMap::new();
    for (top_level_key, series) in source {
        let mut other = 0.;
        let mut these = BTreeMap::new();

        for (name, value) in series {
            let fraction = sums.get(&name).unwrap() / total;
            if fraction > threshold {
                these.insert(name, value);
            } else {
                other += value;
            }
        }

        these.insert("Other".to_string(), other);
        result.insert(top_level_key, these);
    }

    result
}

fn translate_hulls_1d<T: Copy>(
    source: &BTreeMap<TypeID, T>,
) -> Result<BTreeMap<String, T>, TypeError> {
    let ids: Vec<TypeID> = source.keys().copied().collect();
    let types = TypeDB::load_types(&ids)?;

    let mut result = BTreeMap::new();
    for (t, value) in source {
        if let Some(the_type) = types.get(t).unwrap() {
            result.insert(the_type.name.clone(), *value);
        } else {
            return Err(TypeError::NothingMatched);
        }
    }

    Ok(result)
}

fn translate_hulls_2d<K: Copy + Ord, T: Copy>(
    source: &BTreeMap<K, BTreeMap<TypeID, T>>,
) -> Result<BTreeMap<K, BTreeMap<String, T>>, TypeError> {
    let mut result = BTreeMap::new();
    for (k, value) in source {
        result.insert(*k, translate_hulls_1d(value)?);
    }
    Ok(result)
}

struct Queries {}

impl Queries {
    async fn fleet_seconds_by_character_by_month(
        db: &crate::DB,
    ) -> Result<BTreeMap<YearMonth, BTreeMap<i64, f64>>, sqlx::Error> {
        #[derive(sqlx::FromRow)]
        struct Result {
            yearmonth: String,
            character_id: i64,
            time_in_fleet: i64,
        }

        let res: Vec<Result> = sqlx::query_as(concat!(
            "
            SELECT
                ",
            year_month!(from_unixtime!("first_seen")),
            " yearmonth,
                character_id,
                CAST(SUM(last_seen - first_seen) AS SIGNED) time_in_fleet
            FROM fleet_activity
            GROUP BY 1, 2
        "
        ))
        .fetch_all(db)
        .await?;

        let mut result = BTreeMap::new();
        for row in res {
            result
                .entry(YearMonth::parse(&row.yearmonth))
                .or_insert_with(BTreeMap::new)
                .insert(row.character_id, row.time_in_fleet as f64);
        }

        Ok(result)
    }

    async fn fleet_seconds_by_hull_by_month(
        db: &crate::DB,
    ) -> Result<BTreeMap<YearMonth, BTreeMap<TypeID, f64>>, sqlx::Error> {
        #[derive(sqlx::FromRow)]
        struct Result {
            yearmonth: String,
            hull: i64,
            time_in_fleet: i64,
        }

        let res: Vec<Result> = sqlx::query_as(concat!(
            "
            SELECT
                ",
            year_month!(from_unixtime!("first_seen")),
            " yearmonth,
                hull,
                CAST(SUM(last_seen - first_seen) AS SIGNED) time_in_fleet
            FROM fleet_activity
            GROUP BY 1, 2
        "
        ))
        .fetch_all(db)
        .await?;

        let mut result = BTreeMap::new();
        for row in res {
            result
                .entry(YearMonth::parse(&row.yearmonth))
                .or_insert_with(BTreeMap::new)
                .insert(row.hull as TypeID, row.time_in_fleet as f64);
        }

        Ok(result)
    }

    async fn xes_by_hull_by_month(
        db: &crate::DB,
    ) -> Result<BTreeMap<YearMonth, BTreeMap<TypeID, f64>>, sqlx::Error> {
        #[derive(sqlx::FromRow)]
        struct Result {
            yearmonth: String,
            hull: i64,
            x_count: i64,
        }

        let res: Vec<Result> = sqlx::query_as(concat!(
            "
            SELECT
                ",
            year_month!(from_unixtime!("logged_at")),
            " yearmonth,
                hull,
                COUNT(DISTINCT character_id) x_count
            FROM fit_history
            JOIN fitting ON fit_history.fit_id=fitting.id
            GROUP BY 1, 2
        "
        ))
        .fetch_all(db)
        .await?;

        let mut result = BTreeMap::new();
        for row in res {
            result
                .entry(YearMonth::parse(&row.yearmonth))
                .or_insert_with(BTreeMap::new)
                .insert(row.hull as TypeID, row.x_count as f64);
        }

        Ok(result)
    }

    async fn xes_by_hull_28d(db: &crate::DB) -> Result<BTreeMap<TypeID, f64>, sqlx::Error> {
        let ago_28d = chrono::Utc::now().timestamp() - (28 * 86400);

        #[derive(sqlx::FromRow)]
        struct Result {
            hull: TypeID,
            x_count: i64,
        }
        let res: Vec<Result> = sqlx::query_as(concat!(
            "
            SELECT
                hull,
                COUNT(DISTINCT character_id) x_count
            FROM fit_history
            JOIN fitting ON fit_history.fit_id=fitting.id
            WHERE logged_at > ?
            GROUP BY 1
            "
        ))
        .bind(ago_28d)
        .fetch_all(db)
        .await?;

        Ok(res
            .into_iter()
            .map(|row| (row.hull as TypeID, row.x_count as f64))
            .collect())
    }

    async fn fleet_seconds_by_hull_28d(
        db: &crate::DB,
    ) -> Result<BTreeMap<TypeID, f64>, sqlx::Error> {
        let ago_28d = chrono::Utc::now().timestamp() - (28 * 86400);

        #[derive(sqlx::FromRow)]
        struct Result {
            hull: TypeID,
            fleet_seconds: i64,
        }
        let res: Vec<Result> = sqlx::query_as(concat!(
            "
            SELECT
                hull,
                CAST(SUM(last_seen - first_seen) AS SIGNED) fleet_seconds
            FROM fleet_activity
            WHERE first_seen > ?
            GROUP BY 1
        "
        ))
        .bind(ago_28d)
        .fetch_all(db)
        .await?;

        Ok(res
            .into_iter()
            .map(|row| (row.hull as TypeID, row.fleet_seconds as f64))
            .collect())
    }
}

struct Displayer {}
impl Displayer {
    fn build_fleet_seconds_by_hull_by_month(
        source: &BTreeMap<YearMonth, BTreeMap<TypeID, f64>>,
    ) -> Result<BTreeMap<YearMonth, BTreeMap<String, f64>>, Madness> {
        Ok(filter_into_other_2d(translate_hulls_2d(source)?, 0.01))
    }

    fn build_xes_by_hull_by_month(
        source: &BTreeMap<YearMonth, BTreeMap<TypeID, f64>>,
    ) -> Result<BTreeMap<YearMonth, BTreeMap<String, f64>>, Madness> {
        Ok(filter_into_other_2d(translate_hulls_2d(source)?, 0.01))
    }

    fn build_fleet_seconds_by_month(
        source: &BTreeMap<YearMonth, BTreeMap<TypeID, f64>>,
    ) -> BTreeMap<YearMonth, f64> {
        source
            .iter()
            .map(|(month, values)| (*month, values.values().sum()))
            .collect()
    }

    fn build_pilots_by_month(
        source: &BTreeMap<YearMonth, BTreeMap<i64, f64>>,
    ) -> BTreeMap<YearMonth, f64> {
        source
            .iter()
            .map(|(month, values)| (*month, values.len() as f64))
            .collect()
    }

    fn build_xes_by_hull_28d(
        source: &BTreeMap<TypeID, f64>,
    ) -> Result<BTreeMap<String, f64>, Madness> {
        Ok(filter_into_other_1d(translate_hulls_1d(source)?, 0.01))
    }

    fn build_fleet_seconds_by_hull_28d(
        source: &BTreeMap<TypeID, f64>,
    ) -> Result<BTreeMap<String, f64>, Madness> {
        Ok(filter_into_other_1d(translate_hulls_1d(source)?, 0.01))
    }

    fn build_x_vs_time_by_hull_28d(
        source_x: &BTreeMap<TypeID, f64>,
        source_time: &BTreeMap<TypeID, f64>,
    ) -> Result<BTreeMap<String, BTreeMap<&'static str, f64>>, Madness> {
        let sum_x: f64 = source_x.values().sum();
        let sum_time: f64 = source_time.values().sum();
        let translated_x = filter_into_other_1d(translate_hulls_1d(source_x)?, 0.01);
        let translated_time = translate_hulls_1d(source_time)?;

        let mut result = BTreeMap::new();
        for (hull, x_count) in translated_x {
            let mut h = BTreeMap::new();
            h.insert("X", x_count / sum_x);
            h.insert("Time", 0.);
            result.insert(hull, h);
        }

        for (hull, time) in translated_time {
            let mut entry = result.get_mut(&hull);
            if entry.is_none() {
                entry = result.get_mut("Other");
            }
            let entry = entry.unwrap(); // filter_into_other guarantees "Other" exists
            *entry.get_mut("Time").unwrap() += time / sum_time;
        }

        Ok(result)
    }

    fn build_time_spent_in_fleet_by_month(
        source: &BTreeMap<YearMonth, BTreeMap<i64, f64>>,
    ) -> BTreeMap<YearMonth, BTreeMap<&'static str, f64>> {
        let mut result = BTreeMap::new();
        for (month, pilots) in source {
            let mut this_month = BTreeMap::new();
            for time_in_fleet in pilots.values() {
                let bucket = match *time_in_fleet {
                    t if t <= (3600. * 1.) => "a. <1h",
                    t if t <= (3600. * 5.) => "b. 1-5h",
                    t if t <= (3600. * 15.) => "c. 5-15h",
                    t if t <= (3600. * 40.) => "d. 15-40h",
                    _ => "e. 40h+",
                };
                *this_month.entry(bucket).or_default() += 1.;
            }
            result.insert(*month, this_month);
        }
        result
    }
}

#[derive(Serialize)]
struct StatsResponse {
    fleet_seconds_by_hull_by_month: BTreeMap<YearMonth, BTreeMap<String, f64>>,
    xes_by_hull_by_month: BTreeMap<YearMonth, BTreeMap<String, f64>>,
    fleet_seconds_by_month: BTreeMap<YearMonth, f64>,
    pilots_by_month: BTreeMap<YearMonth, f64>,
    xes_by_hull_28d: BTreeMap<String, f64>,
    fleet_seconds_by_hull_28d: BTreeMap<String, f64>,
    x_vs_time_by_hull_28d: BTreeMap<String, BTreeMap<&'static str, f64>>,
    time_spent_in_fleet_by_month: BTreeMap<YearMonth, BTreeMap<&'static str, f64>>,
}

#[get("/api/stats")]
async fn statistics(
    app: &rocket::State<Application>,
    account: AuthenticatedAccount,
) -> Result<Json<StatsResponse>, Madness> {
    account.require_access("stats-view")?;

    let seconds_by_character_month =
        Queries::fleet_seconds_by_character_by_month(app.get_db()).await?;
    let seconds_by_hull_month = Queries::fleet_seconds_by_hull_by_month(app.get_db()).await?;
    let xes_by_hull_month = Queries::xes_by_hull_by_month(app.get_db()).await?;
    let xes_by_hull_28d = Queries::xes_by_hull_28d(app.get_db()).await?;
    let seconds_by_hull_28d = Queries::fleet_seconds_by_hull_28d(app.get_db()).await?;

    Ok(Json(StatsResponse {
        fleet_seconds_by_hull_by_month: Displayer::build_fleet_seconds_by_hull_by_month(
            &seconds_by_hull_month,
        )?,
        xes_by_hull_by_month: Displayer::build_xes_by_hull_by_month(&xes_by_hull_month)?,
        fleet_seconds_by_month: Displayer::build_fleet_seconds_by_month(&seconds_by_hull_month),
        pilots_by_month: Displayer::build_pilots_by_month(&seconds_by_character_month),
        xes_by_hull_28d: Displayer::build_xes_by_hull_28d(&xes_by_hull_28d)?,
        fleet_seconds_by_hull_28d: Displayer::build_fleet_seconds_by_hull_28d(
            &seconds_by_hull_28d,
        )?,
        x_vs_time_by_hull_28d: Displayer::build_x_vs_time_by_hull_28d(
            &xes_by_hull_28d,
            &seconds_by_hull_28d,
        )?,
        time_spent_in_fleet_by_month: Displayer::build_time_spent_in_fleet_by_month(
            &seconds_by_character_month,
        ),
    }))
}

pub fn routes() -> Vec<rocket::Route> {
    routes![statistics]
}
